Lib/subprocess.py

```diff
@@ -809,6 +822,41 @@ def __init__(self, args, bufsize=-1, executable=None,
             else:
                 line_buffering = False
 
+        gids = []
+        if groups is not None:
+            if isinstance(groups, str):
+                gids = [grp.getgrnam(groups).gr_gid]
+            elif isinstance(groups, int):
+                gids = [groups]
+            elif isinstance(groups, list):
+                for group in groups:
+                    if isinstance(group, str):
+                        gids.append(grp.getgrnam(group).gr_gid)
+                    elif isinstance(group, int):
+                        gids.append(group)
+                    else:
+                        raise TypeError("Group must be a string or integer")
+            else:
+                raise TypeError("Group must be a list, a string or an integer")
+
+            # make sure that the gids are all positive here so we can do less
+            # checking in the C code
+            for gid in gids:
+                if gid < 0:
+                    raise ValueError("Group ID cannot be negative")
```

Comment:
on integer based value errors like this it can be helpful for people debugging problems to have the value.  include in in the message (here and below).

 `raise ValueError(f"Group ID cannot be negative, got {gid}.")`